js/src/scrollspy.js

```diff
@@ -85,146 +84,163 @@ class ScrollSpy extends BaseComponent {
+  _getRootMargin() {
+    return this._config.offset ? `${this._config.offset}px 0px 0px` : this._config.rootMargin
+  }
 
-    this._clear()
+  _initializeTargetsAndObservables() {
+    this._targetLinks = new Map()
+    this._observableSections = new Map()
 
-    const queries = SELECTOR_LINK_ITEMS.split(',')
-      .map(selector => `${selector}[data-bs-target="${target}"],${selector}[href="${target}"]`)
+    const targetLinks = SelectorEngine.find(SELECTOR_TARGET_LINKS, this._config.target)
 
-    const link = SelectorEngine.findOne(queries.join(','), this._config.target)
+    for (const anchor of targetLinks) {
+      // ensure that the anchor has an id and is not disabled
+      if (!anchor.hash || isDisabled(anchor)) {
+        continue
+      }
 
-    link.classList.add(CLASS_NAME_ACTIVE)
-    if (link.classList.contains(CLASS_NAME_DROPDOWN_ITEM)) {
-      SelectorEngine.findOne(SELECTOR_DROPDOWN_TOGGLE, link.closest(SELECTOR_DROPDOWN))
-        .classList.add(CLASS_NAME_ACTIVE)
-    } else {
-      for (const listGroup of SelectorEngine.parents(link, SELECTOR_NAV_LIST_GROUP)) {
-        // Set triggered links parents as active
-        // With both <ul> and <nav> markup a parent is the previous sibling of any nav ancestor
-        for (const item of SelectorEngine.prev(listGroup, `${SELECTOR_NAV_LINKS}, ${SELECTOR_LIST_ITEMS}`)) {
-          item.classList.add(CLASS_NAME_ACTIVE)
-        }
+      const observableSection = SelectorEngine.findOne(anchor.hash, this._element)
 
-        // Handle special case when .nav-link is inside .nav-item
-        for (const navItem of SelectorEngine.prev(listGroup, SELECTOR_NAV_ITEMS)) {
-          for (const item of SelectorEngine.children(navItem, SELECTOR_NAV_LINKS)) {
-            item.classList.add(CLASS_NAME_ACTIVE)
-          }
-        }
+      if (observableSection) {
+        this._targetLinks.set(anchor.hash, anchor)
+        this._observableSections.set(anchor.hash, observableSection)
       }
     }
+  }
+
+  _process(target) {
+    if (this._activeTarget === target) {
+      return
+    }
 
-    EventHandler.trigger(this._scrollElement, EVENT_ACTIVATE, {
+    this._clearActiveClass(this._config.target)
+
+    this._activeTarget = target
+
+    target.classList.add(CLASS_NAME_ACTIVE)
+    this._activateParents(target)
+
+    EventHandler.trigger(this._element, EVENT_ACTIVATE, {
       relatedTarget: target
     })
   }
 
-  _clear() {
-    const activeNodes = SelectorEngine.find(SELECTOR_LINK_ITEMS, this._config.target)
-      .filter(node => node.classList.contains(CLASS_NAME_ACTIVE))
+  _activateParents(target) {
+    // Activate dropdown parents
+    if (target.classList.contains(CLASS_NAME_DROPDOWN_ITEM)) {
+      SelectorEngine.findOne(SELECTOR_DROPDOWN_TOGGLE, target.closest(SELECTOR_DROPDOWN))
+        .classList.add(CLASS_NAME_ACTIVE)
+      return
+    }
+
+    for (const listGroup of SelectorEngine.parents(target, SELECTOR_NAV_LIST_GROUP)) {
```

Comment:
Might be worth moving the for of elements to variables here and below; they should be inlined when minified anyway :)